### ISSUE TITLE
How to declare $dynamoDbIndexKeys

### ISSUE THREAD

**Issue by vanquang9387** [baopham/laravel-dynamodb#35]
Hi, I have this table:

```
            'AttributeDefinitions'  => [
                ['AttributeName' => 'parent_id', 'AttributeType' => 'S'],
                ['AttributeName' => 'id', 'AttributeType' => 'S'],
                ['AttributeName' => 'sort_no', 'AttributeType' => 'N']
            ],
            'KeySchema'             => [
                ['AttributeName' => 'parent_id', 'KeyType' => 'HASH'],
                ['AttributeName' => 'id', 'KeyType' => 'RANGE'],
            ],
            'LocalSecondaryIndexes' => [
                [
                    'IndexName'  => 'idx_category_sort',
                    'KeySchema'  => [
                        ['AttributeName' => 'parent_id', 'KeyType' => 'HASH'],
                        ['AttributeName' => 'sort_no', 'KeyType' => 'RANGE']
                    ],
                    'Projection' => [
                        'ProjectionType'   => 'INCLUDE',
                        'NonKeyAttributes' => ['name', 'topics_count', 'state']
                    ]
                ]
            ],
```

How can I declare `$dynamoDbIndexKeys` with this table?
Thank you.

**Comment by baopham** [comment]
I have not tried local secondary indexes. But looks like you're using composite key. Please take a look at the test:

https://github.com/baopham/laravel-dynamodb/blob/282bfeb5e179cacea244453fd7e6a0f0dc660580/tests/DynamoDbCompositeModelTest.php#L175

The schema for the test:
https://github.com/baopham/laravel-dynamodb/blob/282bfeb5e179cacea244453fd7e6a0f0dc660580/dynamodb_local_schema.js#L48

**Comment by vanquang9387** [comment]
@baopham In `composite_test_model`, it looks like you're using only hash key for `count_index`:

```
    GlobalSecondaryIndexes: [ // optional (list of GlobalSecondaryIndex)
        {
            IndexName: 'count_index',
            KeySchema: [
                { // Required HASH type attribute
                    AttributeName: 'count',
                    KeyType: 'HASH',
                }
            ],
            Projection: { // attributes to project into the index
                ProjectionType: 'ALL', // (ALL | KEYS_ONLY | INCLUDE)
            },
            ProvisionedThroughput: { // throughput to provision to the index
                ReadCapacityUnits: 1,
                WriteCapacityUnits: 1,
            },
        },
        // ... more global secondary indexes ...
    ]
```
```
    protected $dynamoDbIndexKeys = [
        'count' => 'count_index',
    ];
```

How about `$dynamoDbIndexKeys` in case of using composit key for global secondary indexes or local secondary indexes?

**Comment by baopham** [comment]
The full example is:

https://github.com/baopham/laravel-dynamodb/blob/282bfeb5e179cacea244453fd7e6a0f0dc660580/tests/DynamoDbCompositeModelTest.php#L181-L185: 

```
    protected $compositeKey = ['id', 'id2'];
    protected $dynamoDbIndexKeys = [
        'count' => 'count_index',
    ];
```

So have you tried:

```
    protected $compositeKey = ['parent_id', 'id'];
    protected $dynamoDbIndexKeys = [
        'parent_id' => 'idx_category_sort',
    ];
```

I have not tried schema similar to you but try if that works. Otherwise, PR is welcome.

**Comment by vanquang9387** [comment]
@baopham I tried but it didn't work.
```
$this->where('parent_id', '0')->where('flg_state', 1)->get();
```
gives this:
$op: 
```
"Query"
```
$query:
```
  "TableName" => "categories"
  "IndexName" => "idx_category_sort"
  "KeyConditions" => array:2 [▼
    "parent_id" => array:2 [▼
      "AttributeValueList" => array:1 [▼
        0 => array:1 [▼
          "S" => "0"
        ]
      ]
      "ComparisonOperator" => "EQ"
    ]
    "flg_state" => array:2 [▼
      "AttributeValueList" => array:1 [▼
        0 => array:1 [▼
          "N" => "1"
        ]
      ]
      "ComparisonOperator" => "EQ"
    ]
  ]
  "ScanFilter" => array:2 [▼
    "parent_id" => array:2 [▼
      "AttributeValueList" => array:1 [▼
        0 => array:1 [▼
          "S" => "0"
        ]
      ]
      "ComparisonOperator" => "EQ"
    ]
    "flg_state" => array:2 [▼
      "AttributeValueList" => array:1 [▼
        0 => array:1 [▼
          "N" => "1"
        ]
      ]
      "ComparisonOperator" => "EQ"
    ]
  ]
```
which return error: `Query condition missed key schema element`
because of this line: [`$query['KeyConditions'] = $this->where;`](https://github.com/baopham/laravel-dynamodb/blob/1aad449cd9a22ccf67d2ed0895b3898016a1688b/src/DynamoDbQueryBuilder.php#L245) will set all conditions (including non-key attribute `flg_state`) in KeyConditions.

Moreover, with `Query` op, `ScanFilter` is ignored, `QueryFilter` or `FilterExpression` should be used instead. Right?

**Comment by baopham** [comment]
So hopefully this can be closed now?

**Comment by vanquang9387** [comment]
@baopham In fact, https://github.com/baopham/laravel-dynamodb/pull/36 only fix `$query['KeyConditions']` contains non-key attributes. I still have no idea with `$dynamoDbIndexKeys` for secondary index having composite key.

**Comment by vanquang9387** [comment]
@baopham This issue can be trully closed with #42
